File: State/cpp-source/TrafficLight.cpp
//
// Created by Pavel Akhtyamov on 2019-04-23.
//

#include "TrafficLight.h"
#include "states/Red.h"

#include <iostream>

TrafficLight::TrafficLight(): direction_(TO_GREEN) {
  state_ = std::make_unique<Red>();
}

void TrafficLight::SetState(StatePointer state) {
  state_ = std::move(state);
}

Direction TrafficLight::GetDirection() {
  return direction_;
}

void TrafficLight::SetDirection(Direction direction) {
  direction_ = direction;
}

void TrafficLight::ChangeColor() {
  state_->Switch(this);
}

void TrafficLight::PrintState() {
  std::cout << "State" << std::endl;
  std::cout << "Direction: " << direction_ << std::endl;
  std::cout << "Color: " << state_->GetStringState() << std::endl;
}

File: JavaSource/src/main/java/ru/akhcheck/patterns/state/TrafficLight.java
package ru.akhcheck.patterns.state;

import ru.akhcheck.patterns.state.states.Red;
import ru.akhcheck.patterns.state.states.State;

/**
 * Хозяин состояния. Про цвета не знает ничего.
 *
 * <p>Обрати внимание, чего здесь нет: ни одного условия вида «если сейчас
 * красный, то дальше жёлтый». Такие условия живут в самих состояниях, и
 * добавление нового цвета не трогает этот класс.
 */
public class TrafficLight {
    private State state = new Red();
    private Direction direction = Direction.TO_GREEN;

    public void setState(State state) {
        this.state = state;
    }

    public Direction getDirection() {
        return direction;
    }

    public void setDirection(Direction direction) {
        this.direction = direction;
    }

    public void changeColor() {
        state.switchState(this);
    }

    public void printState() {
        System.out.println(state.getStringState());
    }
}
